locksmith/src/controllers/hookController.ts

```diff
@@ -0,0 +1,192 @@
+export async function validHubIntent(hub: z.infer<typeof Hub>) {
+  const challenge = crypto.randomBytes(20).toString('hex')
+  const result = await fetch(hub.callback, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({
+      hub: {
+        ...hub,
+        challenge,
+      },
+    }),
+  })
+
+  const json = await result.json()
+
+  if (!result.ok) {
+    throw new Error("Subscriber didn't confirm intent.")
+  }
+
+  if (json.challenge !== challenge) {
+    throw new Error('Challenge did not match')
+  }
+
+  if (json.topic !== hub.topic) {
+    throw new Error('Topic did not match')
+  }
+
+  if (json.callback !== hub.callback) {
+    throw new Error('Callback did not match')
+  }
+
+  return hub
+}
+
+export function getSupportedNetwork(network: string) {
+  if (!networks[network]) {
+    logger.error(`Unsupported network: ${network}`)
+    return
+  }
+  return networks[network]
+}
+
+export async function subscriptionHandler(req: Request, res: Response) {
+  try {
+    // We check the hub schema here to make sure the subscriber is sending us the correct data.
+    const hub = await Hub.parseAsync(req.body.hub)
+    // We check the network here to make sure the subscriber is sending to the right network endpoint.
+    const network = getSupportedNetwork(req.params.network)
+    if (!network) {
+      return res.status(400).send('Unsupported network')
+    }
+    // We send the accepted request to the subscriber and then validate the intent of the subscriber as well as persist the subscription.
+    res.status(202).send('Accepted')
+    return subscribe(hub, req.params)
+  } catch (error) {
+    logger.error(error.message)
+    // We respond with the error if we cannot subscribe or there is an error in the subscriber request.
+    return res.status(500).send({
+      hub: {
+        mode: req.body.hub.mode,
+        topic: req.body.hub.topic,
+        reason: error.message,
+      },
+    })
+  }
+}
+
+export function publisherHandler(req: Request, res: Response) {
+  try {
+    const network = getSupportedNetwork(req.params.network)
+    if (!network) {
+      return res.status(400).send('Unsupported network')
+    }
+
+    const url = new URL(req.originalUrl, `${req.protocol}://${req.hostname}`)
+    const links = [
+      {
+        rel: 'self',
+        href: url.toString(),
+      },
+      {
+        rel: 'hub',
+        href: url.toString(),
+      },
+    ]
+    res.setHeader(
+      'Link',
+      links.map((item) => `<${item.href}>; rel="${item.rel}"`)
+    )
+
+    const htmlResponse = `
+      <!DOCTYPE html>
+      <html>
+        <head>
+          ${links
+            .map((item) => html`<link rel="${item.rel}" href="${item.href}" />`)
+            .join('')}
+        </head>
+        <body>
+           To subscribe to our topic, make a POST request to this endpoint with the valid hub payload.
```

Comment:
I think a link to the Websub spec might be helpful here.

Reply:
okay will link it.